tests/v2/test_waku_rln_relay_onchain.nim

```diff
@@ -0,0 +1,263 @@
+
+# contains rln-relay tests that require interaction with Ganache i.e., onchain tests
+{.used.}
+
+import
+  std/options, sequtils, times,
+  testutils/unittests, chronos, chronicles, stint, web3, json,
+  stew/byteutils, stew/shims/net as stewNet,
+  libp2p/crypto/crypto,
+  ../../waku/v2/protocol/waku_rln_relay/[rln, waku_rln_relay_utils,
+      waku_rln_relay_types, rln_relay_contract],
+  ../../waku/v2/node/wakunode2,
+  ../test_helpers,
+  ./test_utils
+
+const RLNRELAY_PUBSUB_TOPIC = "waku/2/rlnrelay/proto"
+const RLNRELAY_CONTENT_TOPIC = "waku/2/rlnrelay/proto"
+
+#  contract ABI
+contract(MembershipContract):
+  proc register(pubkey: Uint256) # external payable
+  proc MemberRegistered(pubkey: Uint256, index: Uint256) {.event.}
+  # proc registerBatch(pubkeys: seq[Uint256]) # external payable
+  # proc withdraw(secret: Uint256, pubkeyIndex: Uint256, receiver: Address)
+  # proc withdrawBatch( secrets: seq[Uint256], pubkeyIndex: seq[Uint256], receiver: seq[Address])
+
+#  a util function used for testing purposes
+#  it deploys membership contract on Ganache (or any Eth client available on ETH_CLIENT address)
+#  must be edited if used for a different contract than membership contract
+proc uploadRLNContract*(ethClientAddress: string): Future[Address] {.async.} =
+  let web3 = await newWeb3(ethClientAddress)
+  debug "web3 connected to", ethClientAddress
+
+  # fetch the list of registered accounts
+  let accounts = await web3.provider.eth_accounts()
+  web3.defaultAccount = accounts[1]
+  let add = web3.defaultAccount
+  debug "contract deployer account address ", add
+
+  var balance = await web3.provider.eth_getBalance(web3.defaultAccount, "latest")
+  debug "Initial account balance: ", balance
+
+  # deploy the poseidon hash contract and gets its address
+  let
+    hasherReceipt = await web3.deployContract(POSEIDON_HASHER_CODE)
+    hasherAddress = hasherReceipt.contractAddress.get
+  debug "hasher address: ", hasherAddress
+
+
+  # encode membership contract inputs to 32 bytes zero-padded
+  let
+    membershipFeeEncoded = encode(MEMBERSHIP_FEE).data
+    depthEncoded = encode(MERKLE_TREE_DEPTH.u256).data
```

Comment:
The registration fee is supposed to be set so that an attacker cannot exhaust a large portion of the leaves for free. I suspect such an entity with lots of coins is a reasonable adversarial assumption. An attacker like that needs to possess almost 2^20*30 USD = 30M USD! (see this issue for the registration cost https://github.com/vacp2p/research/issues/56)

Reply:
At the termination of the DoS attack, the attacker will lose half of his deposit so he would not be able to recover his entire money back. In the preceding example, he will burn 15M USD!

Reply:
>As we already discussed https://github.com/vacp2p/rfc/pull/484#pullrequestreview-879074788, it may be worth brainstorm on alternative constructions which provide a good decentralized/efficiency tradeoff.Nevertheless, the ability to have a more decentralized group management + good incentives would be a nice idea regardless.